File: archiso/airootfs/usr/share/churros/preferences/widgets/slider_row.py
import gi

gi.require_version("Gtk", "4.0")

from gi.repository import Gtk

from widgets.row import Row


class SliderRow(Row):

    def __init__(
        self,
        title,
        icon=None,
        subtitle=None,
        minimum=0,
        maximum=100,
        step=1,
        value=0,
        callback=None
    ):

        self.scale = Gtk.Scale.new_with_range(
            Gtk.Orientation.HORIZONTAL,
            minimum,
            maximum,
            step
        )

        self.scale.set_draw_value(False)

        self.scale.set_hexpand(True)

        self.scale.set_value(
            value
        )

        self.scale.set_size_request(
            180,
            -1
        )

        if callback is not None:

            self.scale.connect(
                "value-changed",
                callback
            )

        super().__init__(
            title=title,
            subtitle=subtitle,
            icon=icon,
            suffix=self.scale
        )

    def get_value(self):

        return self.scale.get_value()

    def set_value(
        self,
        value
    ):

        self.scale.set_value(
            value
        )

File: rust/preferences/src/widgets/slider_row.rs
// ==========================================
// SliderRow — fila con Gtk::Scale
// (equivalente a widgets/slider_row.py)
// ==========================================

use gtk::prelude::*;

use crate::widgets::row::Row;

pub struct SliderRow {
    pub row: Row,
    pub scale: gtk::Scale,
}

impl SliderRow {
    pub fn new(
        title: &str,
        icon: Option<&str>,
        subtitle: Option<&str>,
        minimum: f64,
        maximum: f64,
        step: f64,
        value: f64,
        callback: Option<Box<dyn Fn(f64)>>,
    ) -> Self {
        let scale = gtk::Scale::with_range(gtk::Orientation::Horizontal, minimum, maximum, step);
        scale.set_draw_value(false);
        scale.set_hexpand(true);
        scale.set_value(value);
        scale.set_size_request(180, -1);

        if let Some(cb) = callback {
            scale.connect_value_changed(move |scale| {
                cb(scale.value());
            });
        }

        let scale_widget: &gtk::Widget = scale.upcast_ref();
        let row = Row::new(title, subtitle, icon, None, Some(scale_widget), None);

        Self { row, scale }
    }

    #[allow(dead_code)]
    pub fn widget(&self) -> &gtk::Button {
        self.row.widget()
    }

    pub fn get_value(&self) -> f64 {
        self.scale.value()
    }

    #[allow(dead_code)]
    pub fn set_value(&self, value: f64) {
        self.scale.set_value(value);
    }
}
